Prevent `rmake.rs` from using unstable features, and fix 3 run-make tests that currently do

jieyouxu opened the pull request:
Addresses (mostly) #137532.
Follow-up to #137373.

### Summary

- Fix 3 run-make tests that currently use unstable features:
    1. `tests/run-make/issue-107495-archive-permissions/rmake.rs` uses `#![feature(rustc_private)]` for `libc` on `unix`, but `run_make_support` already exports `libc`, so just use that.
    2. `tests/run-make/cross-lang-lto/rmake.rs` uses `#![feature(path_file_prefix)]` for convenience, replaced with similar filename prefix logic.
    3. `tests/run-make/broken-pipe-no-ice/rmake.rs` uses `#![feature(anonymous_pipe)]` for anonymous pipes. This is more complicated[^race-condition], and I decided to temporarily introduce a dependency on [`os_pipe`] before std's `anonymous_pipe` library feature is stabilized[^pipe-stab]. I left a FIXME tracked by #137532 to make the switch once `anonymous_pipe` stabilizes and reaches beta.
- Use `RUSTC_BOOTSTRAP=-1` when building `rmake.rs` to have the stage 0 rustc reject any unstable features used in `rmake.rs`.

- The requirement that `rmake.rs` may not use any unstable features is now documented in rustc-dev-guide.
- This PR does not impose `RUSTC_BOOTSTRAP=-1` when building `run-make-support`, but I suppose we could.

r? @Kobzol

try-job: x86_64-msvc-1
try-job: x86_64-mingw-1

[`os_pipe`]: https://github.com/oconnor663/os_pipe.rs

[^race-condition]: We can't just try to spawn `rustc` and immediate close the stderr handle because of race condition, as there's no guarantee `rustc` will not try to print to stderr before the handle gets closed.
[^pipe-stab]: In-progress stabilization PR over at https://github.com/rust-lang/rust/pull/135822.

jieyouxu committed:
Fix `issue-107495-archive-permissions` to not rely on `rustc_private`

jieyouxu committed:
Fix `cross-lang-lto` to not use `path_file_prefix`

jieyouxu committed:
Temporarily depend on `os_pipe` in `run-make-support` and re-export it

For `broken-pipe-no-ice` until std `anonymous_pipe` stabilizes.

jieyouxu committed:
Fix `broken-pipe-no-ice` to not depend on unstable `anonymous_pipe` feature

jieyouxu committed:
Prevent `rmake.rs` from using any nightly/unstable features

jieyouxu committed:
Document that `rmake.rs`/`run-make-support` may not use unstable features

jieyouxu commented:
@rustbot blocked (on #137373)

Kobzol reviewed (commented):
Thanks! I think that we should do the same for `run-make-support` (disable unstable features), but it's probably not trivial given it is now defined by a macro. Let me know once this is rebased after my PR is merged.

jieyouxu commented on tests/run-make/cross-lang-lto/rmake.rs:
```diff
@@ -92,10 +90,17 @@ fn check_bitcode(instructions: LibBuild) {
-        has_prefix(path, instructions.output.file_prefix().unwrap().to_str().unwrap())
+    let objects = shallow_find_files(cwd(), |path| {
+        let mut output_path = instructions.output.clone();
+        output_path.set_extension("");
```
Yeah it is not equivalent, however, this is kinda what we want, because if there's something like `$name$hash.cgu0.o` AFAIK we still want to check that file.

Kobzol commented:
@rustbot label -S-blocked

BoxyUwU commented:
TIP: you can start your PR title with the word "wipe" and then it wont be merged, that way you can ensure it isn't merged while its still based on top of another PR

jieyouxu commented:
> TIP: you can start your PR title with the word "wipe" and then it wont be merged, that way you can ensure it isn't merged while its still based on top of another PR

Fair. It's mostly a remark for myself though 😆

jieyouxu commented:
> I think that we should do the same for `run-make-support` (disable unstable features), but it's probably not trivial given it is now defined by a macro.

I briefly looked at this, doesn't seem worth it to add a special hack to the `ToolBuild` step or add yet another thing to the macro.

So only rebased, no functional changes.
@rustbot ready

Kobzol commented:
Thanks, looks good! You can r=me once CI is green.

jieyouxu commented:
@bors r=Kobzol rollup=maybe

bors commented:
:pushpin: Commit dcc4c6c1bf02ec57a4ed6bc015e85bdd0f9b6f08 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol dcc4c6c1bf02ec57a4ed6bc015e85bdd0f9b6f08 -->
<!-- homu: {"type":"Approved","sha":"dcc4c6c1bf02ec57a4ed6bc015e85bdd0f9b6f08","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jhpratt commented:
@bors r-

https://github.com/rust-lang/rust/pull/138146#issuecomment-2705881581

jieyouxu committed:
Apply `--cfg windows_raw_dylib` for bootstrap tools too

This is already applied for rustc/std tools. This is needed to make
`windows` crates avoid trying to depend on a generated
`windows.0.xx.0.lib`.

jieyouxu commented on Cargo.lock:
```diff
@@ -2526,6 +2526,16 @@ version = "0.2.0"
+checksum = "5ffd2b0a5634335b135d5728d84c5e0fd726954b87111f7506a61c502280d982"
+dependencies = [
+ "libc",
+ "windows-sys 0.59.0",
```
Remark [RAW_DYLIB 1/2]: `os_pipe` depends on `windows-sys` which tries to link in `windows.0.52.0.lib` unless you pass a `--cfg windows_raw_dylib`. This is already what rustc / rustc tool cargo uses. I.e. the comment

```
// By default, windows-rs depends on a native library that doesn't get copied into the
// sysroot. Passing this cfg enables raw-dylib support instead, which makes the native
// library unnecessary. This can be removed when windows-rs enables raw-dylib
// unconditionally.
```

applies here.

jieyouxu commented on src/bootstrap/src/core/builder/cargo.rs:
```diff
@@ -597,7 +597,7 @@ impl Builder<'_> {
         // library unnecessary. This can be removed when windows-rs enables raw-dylib
         // unconditionally.
-        if let Mode::Rustc | Mode::ToolRustc = mode {
+        if let Mode::Rustc | Mode::ToolRustc | Mode::ToolBootstrap = mode {
             rustflags.arg("--cfg=windows_raw_dylib");
         }
```
Remark [RAW_DYLIB 2/2]: here.

jieyouxu commented on src/doc/rustc-dev-guide/src/tests/compiletest.md:
```diff
@@ -415,6 +415,10 @@ Compiletest directives like `//@ only-<target>` or `//@ ignore-<target>` are
 supported in `rmake.rs`, like in UI tests. However, revisions or building
 auxiliary via directives are not currently supported.
 
+`rmake.rs` and `run-make-support` may *not* use any nightly/unstable features,
+as they must be compilable by a stage 0 rustc that may be a beta or even stable
+rustc.
```
Remark: adjusted this to also say that `run-make-support` may not depend on nightly/unstable features.

jieyouxu commented:
This passes on msvc locally, but let's try anyway.
@bors try

Kobzol commented:
You can r=me if the try job passes.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13719561359/job/38380342693)
Build commit: 486bfd6307b46a8c375b80de379e0f1df3440659 (`486bfd6307b46a8c375b80de379e0f1df3440659`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13719561359/job/38380342693"},"merge_sha":"486bfd6307b46a8c375b80de379e0f1df3440659"} -->

jieyouxu commented:
@bors r=Kobzol

bors commented:
:pushpin: Commit c9c572cd927b223d22280ca606086d64d20b733b has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol c9c572cd927b223d22280ca606086d64d20b733b -->
<!-- homu: {"type":"Approved","sha":"c9c572cd927b223d22280ca606086d64d20b733b","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
